File: packages/oauth/oauth-client-expo/ios/Jose.swift
import JOSESwift

class JoseUtil: NSObject {
  private static func headerStringToHeader(_ headerString: String) -> JWSHeader? {
    guard let headerData = headerString.data(using: .utf8) else {
      return nil
    }
    return JWSHeader(headerData)
  }

  private static func payloadStringToPayload(_ payloadString: String) -> Payload? {
    guard let payloadData = payloadString.data(using: .utf8) else {
      return nil
    }
    return Payload(payloadData)
  }

  static func createJwt(header: String, payload: String, jwk: SecKey) throws -> String {
    guard let header = headerStringToHeader(header) else {
      throw ExpoAtprotoOAuthClientError.invalidHeader("could not parse header string")
    }

    guard let payload = payloadStringToPayload(payload) else {
      throw ExpoAtprotoOAuthClientError.invalidPayload("could not parse payload string")
    }

    let signer = Signer(signingAlgorithm: .ES256, key: jwk)

    guard let signer = signer else {
      throw ExpoAtprotoOAuthClientError.nullSigner
    }

    let jws = try JWS(header: header, payload: payload, signer: signer)

    return jws.compactSerializedString
  }

  static func verifyJwt(token: String, jwk: SecKey, options: VerifyOptions) throws -> VerifyResult {
    guard let jws = try? JWS(compactSerialization: token),
          let verifier = Verifier(verifyingAlgorithm: .ES256, key: jwk),
          let validation = try? jws.validate(using: verifier)
    else {
      throw ExpoAtprotoOAuthClientError.invalidJwk
    }

    let header = validation.header
    let payload = String(data: validation.payload.data(), encoding: .utf8)
    guard let payload = payload else {
      throw ExpoAtprotoOAuthClientError.invalidPayload("unable to parse payload")
    }

    var protectedHeader: [String: Any] = [:]
    protectedHeader["alg"] = "ES256"
    if header.jku != nil {
      protectedHeader["jku"] = header.jku?.absoluteString
    }
    if header.kid != nil {
      protectedHeader["kid"] = header.kid
    }
    if header.typ != nil {
      protectedHeader["typ"] = header.typ
    }
    if header.cty != nil {
      protectedHeader["cty"] = header.cty
    }
    if header.crit != nil {
      protectedHeader["crit"] = header.crit
    }

    if let typ = options.typ {
      if header.typ != typ {
        throw ExpoAtprotoOAuthClientError.invalidPayload("typ mismatch")
      }
    }

    let claims = try JSONSerialization.jsonObject(with: validation.payload.data(), options: []) as? [String: Any]

    if let requiredClaims = options.requiredClaims {
      try requiredClaims.forEach { c in
        if claims?[c] == nil {
          throw ExpoAtprotoOAuthClientError.invalidPayload("required claim \(c) missing")
        }
      }
    }

    if let audience = options.audience {
      if claims?["aud"] as? String != audience {
        throw ExpoAtprotoOAuthClientError.invalidPayload("audience mismatch")
      }
    }

    if let subject = options.subject {
      if claims?["sub"] as? String != subject {
        throw ExpoAtprotoOAuthClientError.invalidPayload("subject mismatch")
      }
    }

    if let checkTolerance = options.clockTolerance {
      let now = Date()
      let expiryDate: Date
      if let expiryString = claims?["exp"] as? String {
        let formatter = ISO8601DateFormatter()
        expiryDate = formatter.date(from: expiryString)!
      } else {
        throw ExpoAtprotoOAuthClientError.invalidPayload("expiry missing")
      }
      if expiryDate < now - checkTolerance {
        throw ExpoAtprotoOAuthClientError.invalidPayload("token expired")
      }
    }

    if let maxTokenAge = options.maxTokenAge {
      let now = Date()
      if let expiryString = claims?["exp"] as? String {
        let formatter = ISO8601DateFormatter()
        let expiryDate = formatter.date(from: expiryString)!
        if expiryDate < now - maxTokenAge {
          throw ExpoAtprotoOAuthClientError.invalidPayload("token expired")
        }
      } else {
        throw ExpoAtprotoOAuthClientError.invalidPayload("expiry missing")
      }
    }

    if let issuer = options.issuer {
      if claims?["iss"] as? String != issuer {
        throw ExpoAtprotoOAuthClientError.invalidPayload("issuer mismatch")
      }
    }

    let res = VerifyResult()
    res.payload = payload
    res.protectedHeader = protectedHeader

    return res
  }
}

File: packages/oauth/oauth-client-expo/android/src/main/java/expo/modules/atprotooauthclient/Jose.kt
package expo.modules.atprotooauthclient

import com.nimbusds.jose.JWSHeader
import com.nimbusds.jose.crypto.ECDSASigner
import com.nimbusds.jose.crypto.ECDSAVerifier
import com.nimbusds.jose.jwk.ECKey
import com.nimbusds.jwt.JWTClaimsSet
import com.nimbusds.jwt.SignedJWT
import expo.modules.atprotooauthclient.VerifyOptions
import expo.modules.atprotooauthclient.VerifyResult

class InvalidPayloadException(
  message: String,
) : Exception(message)

class Jose {
  fun createJwt(
    header: String,
    payload: String,
    jwk: ECKey,
  ): String {
    val parsedHeader = JWSHeader.parse(header)
    val parsedPayload = JWTClaimsSet.parse(payload)

    val signer = ECDSASigner(jwk)
    val jwt = SignedJWT(parsedHeader, parsedPayload)
    jwt.sign(signer)

    return jwt.serialize()
  }

  fun verifyJwt(
    token: String,
    jwk: ECKey,
    options: VerifyOptions,
  ): VerifyResult {
    val jwt = SignedJWT.parse(token)
    val verifier = ECDSAVerifier(jwk)

    if (!jwt.verify(verifier)) {
      throw InvalidPayloadException("invalid JWT signature")
    }

    val protectedHeader = emptyMap<String, Any>().toMutableMap()
    protectedHeader["alg"] = jwt.header.algorithm

    jwt.header.getCustomParam("jku")?.let {
      protectedHeader["jku"] = it.toString()
    }
    jwt.header.keyID?.let {
      protectedHeader["kid"] = it
    }
    jwt.header.type?.let {
      protectedHeader["typ"] = it.toString()
    }
    jwt.header.contentType?.let {
      protectedHeader["cty"] = it
    }
    jwt.header.criticalParams?.let {
      protectedHeader["crit"] = it.toList()
    }

    options.typ?.let {
      if (jwt.header.type.toString() != it) {
        throw InvalidPayloadException("typ mismatch")
      }
    }

    val claims = jwt.jwtClaimsSet

    options.requiredClaims?.let { requiredClaims ->
      requiredClaims.forEach { claim ->
        if (!claims.claims.containsKey(claim)) {
          throw InvalidPayloadException("required claim '$claim' missing")
        }
      }
    }

    options.audience?.let {
      if (!claims.audience.contains(it)) {
        throw InvalidPayloadException("audience mismatch")
      }
    }

    options.subject?.let {
      if (claims.subject != it) {
        throw InvalidPayloadException("subject mismatch")
      }
    }

    options.checkTolerance?.let {
      val currentTime = options.currentDate ?: (System.currentTimeMillis() / 1000.0)
      if (claims.issueTime.time / 1000.0 + it < currentTime) {
        throw InvalidPayloadException("token expired")
      }
    }

    options.maxTokenAge?.let {
      val currentTime = options.currentDate ?: (System.currentTimeMillis() / 1000.0)
      if (claims.issueTime.time / 1000.0 + it < currentTime) {
        throw InvalidPayloadException("token expired")
      }
    }

    options.issuer?.let {
      if (claims.issuer != it) {
        throw InvalidPayloadException("issuer mismatch")
      }
    }

    return VerifyResult().apply {
      payload = jwt.payload.toString()
      this.protectedHeader = protectedHeader
    }
  }
}
